[RNMobile] Fix drag mode not being enabled when long-pressing over Shortcode block

fluiddot opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes the drag mode not being enabled due to the use of `PlainText` component in the Shortcode block.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

This PR will address the issue described in https://github.com/wordpress-mobile/gutenberg-mobile/issues/4853, specifically the one related to the Shortcode block. In the future, we could use the same approach for the rest of the affected blocks.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

The fix consists in using the `RichText` component by disabling all formatting options for rendering plain text via the `PlainText` component, similarly as it's done experimentally in the web version ([reference](https://github.com/WordPress/gutenberg/blob/a1e1fdc0e6278457e9f4fc0b31ac6d2095f5450b/packages/block-editor/src/components/plain-text/index.js#L21-L23)). In order to achieve this, I also had to apply the following modifications in the `RichText` component:
- Add `disableSuggestions` prop to be able to disable the suggestion button in the formatting toolbar.
- Add `disableAutocorrection` prop to be able to disable auto-correction settings of the text input. This change also implied updating the native code of `react-native-aztec` package.
- Fix the format types calculation by passing `allowedFormats `value to `useFormatTypes` hook.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

### Drag mode is enabled when long-pressing the content
1. Open a post/page.
2. Add a Shortcode block and add some text to it.
3. Long-press over the content of the Shortcode block.
4. Observe that the drag mode is enabled and that it allows dropping the block in the desired location.

### Default long-press behavior when editing text
1. Open a post/page.
2. Add a Shortcode block and add some text to it.
3. While editing the text, long-press over parts of the text.
4. Observe that the long-press gesture behaves as expected and doesn't enable the drag mode.

### Shortcode block
1. Open a post/page.
2. Add a Shortcode block and add some text to it.
3. Observe that the block styles match the previous version by comparing it with the latest production version of the WordPress app.
4. Observe that, when editing the content of the Shortcode block, there are no formatting buttons.
5. Save the post.
6. Observe that the HTML generated for the Shortcode block is correct by comparing the same output using the web version.

### Check enable/disable suggestions setting in `RichText`
1. Open a post/page.
2. Add a Paragraph block and edit its content.
3. Observe that the Suggestions button (i.e. the button with `@` symbol) is displayed in the formatting toolbar.
4. Add a Shortcode block and edit its content.
5. Observe that the Suggestion button is NOT displayed in the formatting toolbar.

### Check enable/disable Auto-correction setting in `RichText`
1. Open a post/page.
2. Add a Paragraph block and edit its content.
3. Observe that when typing text it's autocorrected (e.g. typing `morninn` is converted to `morning`).
4. Add a Shortcode block and edit its content.
5. Observe that when typing text it's NOT autocorrected.

### Check other blocks using `PlainText`
1. Open a post/page.
2. Add a Search block.
3. Observe that the block works as expected.
4. Add a Contact Info block.
5. Observe that the block works as expected.

## Screenshots or screencast <!-- if applicable -->
Android|iOS
--|--
<video src=https://user-images.githubusercontent.com/14905380/169321953-31126240-a268-4406-8c8f-217d5b8cc41f.mp4>|<video src=https://user-images.githubusercontent.com/14905380/169322048-91d399ec-2969-418d-9ba4-9c2653c19268.mp4>

fluiddot committed:
Add prop for disabling suggestions button

fluiddot committed:
Use allowed formats in format types calculation

fluiddot committed:
Add RichText version to PlainText component

fluiddot committed:
Use experimental version of PlainText in Shortcode block

fluiddot committed:
Add disableAutocorrection prop to RichText

fluiddot committed:
Disable autocorrection in Shortcode block

fluiddot committed:
Update PlainText props in Shortcode block

fluiddot committed:
Use pre as tagName in PlainText

fluiddot committed:
Rename replaceLineBreaks function

fluiddot committed:
Update shortcode block unit tests

geriux commented on packages/block-editor/src/components/plain-text/index.native.js:
```diff
@@ -79,13 +79,65 @@ export default class PlainText extends Component {
+						}
+					} }
+					unstableOnFocus={ onFocus }
+					__unstableMobileNoFocusOnMount={ true }
```
When adding a Shortcode block, due to this prop the block doesn't autofocus, is this needed to avoid some RichText issue?

fluiddot replied:
Thanks for pointing it out 🙇 ! Yep, I also noticed this recently when testing, I finally removed the prop 👍 .

fluiddot commented on packages/react-native-aztec/ios/RNTAztecView/RCTAztecView.swift:
```diff
@@ -31,6 +31,12 @@ class RCTAztecView: Aztec.TextView {
         }
     }
 
+    @objc var disableAutocorrection: Bool = false {
+        didSet {
+            autocorrectionType = disableAutocorrection ? .no : .default
+        }
+    }
```
This logic is similar to the RN's `TextInput` component on iOS ([reference](https://github.com/facebook/react-native/blob/402ad820270fe83e3efc9ee652a6c87de919d60e/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputUtils.mm#L54-L58)).

fluiddot commented on packages/react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecManager.java:
```diff
@@ -617,6 +618,16 @@ public void setShouldDeleteEnter(final ReactAztecText view, boolean shouldDelete
         view.shouldDeleteEnter = shouldDeleteEnter;
     }
 
+    @ReactProp(name = "disableAutocorrection", defaultBoolean = false)
+    public void disableAutocorrection(final ReactAztecText view, boolean disable) {
+        if (disable) {
+            view.setInputType((view.getInputType() & ~InputType.TYPE_TEXT_FLAG_AUTO_CORRECT) | InputType.TYPE_TEXT_FLAG_NO_SUGGESTIONS);
+        }
+        else {
+            view.setInputType((view.getInputType() & ~InputType.TYPE_TEXT_FLAG_NO_SUGGESTIONS) | InputType.TYPE_TEXT_FLAG_AUTO_CORRECT);
+        }
+    }
```
This logic is similar to the RN's `TextInput` component on Android ([reference](https://github.com/facebook/react-native/blob/889578a1420814615c905e36513904d4cbcd77ff/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java#L771-L782)).

fluiddot commented on packages/rich-text/src/component/index.native.js:
```diff
@@ -1249,10 +1253,17 @@ RichText.defaultProps = {
+	const {
+		clientId,
+		identifier,
+		withoutInteractiveFormatting,
+		allowedFormats,
+	} = props;
 	const { formatTypes } = useFormatTypes( {
-		clientId: props.clientId,
-		identifier: props.identifier,
-		withoutInteractiveFormatting: props.withoutInteractiveFormatting,
+		clientId,
+		identifier,
+		withoutInteractiveFormatting,
+		allowedFormats,
```
I found out that the format types weren't taking into account the allowed formats calculated in the `RichText` component:
https://github.com/WordPress/gutenberg/blob/a1e1fdc0e6278457e9f4fc0b31ac6d2095f5450b/packages/block-editor/src/components/rich-text/index.native.js#L194-L198

This was causing that setting the prop `__unstableDisableFormats` doesn't hide the text color format button from the toolbar, as well as preventing hiding other format buttons when using the `allowedFormats` prop.

geriux commented:
Using a block-based theme, RichText sets the text color from the theme as a fallback if none is passed down as a prop:

<kbd><img src="https://user-images.githubusercontent.com/4885740/169310902-da25dbe5-7c88-4699-89b0-9602100d1df7.png" width=240 /></kbd>

I think we could set the default color [here](https://github.com/WordPress/gutenberg/pull/41155/files#diff-592008283ab87fb91b31ecd0ca588945b5ccb8a4408338201a786b0496010eabR24), to avoid this issue, what do you think?

fluiddot committed:
Prevent text input focus when selecting Shortcode block

fluiddot reviewed (commented):
@geriux the PR is ready for review, let me know if you could take a look, thanks 🙇 !

fluiddot commented:
> Using a block-based theme, RichText sets the text color from the theme as a fallback if none is passed down as a prop:
> 
> <img alt="" width="240" src="https://user-images.githubusercontent.com/4885740/169310902-da25dbe5-7c88-4699-89b0-9602100d1df7.png">
> 
> I think we could set the default color [here](https://github.com/WordPress/gutenberg/pull/41155/files#diff-592008283ab87fb91b31ecd0ca588945b5ccb8a4408338201a786b0496010eabR24), to avoid this issue, what do you think?

Oh, good catch! Yeah, I think in this case we should set the text color to prevent it to be overridden by the theme colors 👍 . I'll update it as soon as possible.

fluiddot committed:
Force text color in Shortcode block

fluiddot commented:
> > Using a block-based theme, RichText sets the text color from the theme as a fallback if none is passed down as a prop:
> > <img alt="" width="240" src="https://user-images.githubusercontent.com/4885740/169310902-da25dbe5-7c88-4699-89b0-9602100d1df7.png">
> > I think we could set the default color [here](https://github.com/WordPress/gutenberg/pull/41155/files#diff-592008283ab87fb91b31ecd0ca588945b5ccb8a4408338201a786b0496010eabR24), to avoid this issue, what do you think?
> 
> Oh, good catch! Yeah, I think in this case we should set the text color to prevent it to be overridden by the theme colors 👍 . I'll update it as soon as possible.

This issue has been addressed in https://github.com/WordPress/gutenberg/pull/41155/commits/af35786b0e2fff44462acfd16e5b55b2697dd0eb.

Before|After
--|--
<img src=https://user-images.githubusercontent.com/14905380/169326133-ad37c261-e8ee-44de-b740-f56a29dfc9d4.jpg width=350>|<img src=https://user-images.githubusercontent.com/14905380/169326146-b0ab4580-57a5-4ae8-8d64-26a2d1b55201.jpg width=350>

**NOTE:** I used the `Remote` theme for testing.

geriux commented on packages/block-editor/src/components/plain-text/index.native.js:
```diff
@@ -79,20 +82,74 @@ export default class PlainText extends Component {
+					{ ...forcePlainTextProps }
+					{ ...fontProps }
+					identifier="content"
+					tagName="pre"
```
I think this should come from props, currently, it's fine because only the `Shortcode` block uses it but if we want to use it in other blocks, the `tagName` would be different, what do you think?

fluiddot replied:
Yep, good point. In fact, I think that we could even remove the prop to use the default value `div` ([reference](https://github.com/WordPress/gutenberg/blob/51db4bf888e6b18cf9d18266108114d61070f3ad/packages/rich-text/src/component/index.native.js#L1248)). As far as I checked, it works even better than using `pre` tag 😅 .

Android|iOS
--|--
<video src=https://user-images.githubusercontent.com/14905380/169492308-52798d75-c6d2-4f44-8a42-270e2745924a.mp4>|<video src=https://user-images.githubusercontent.com/14905380/169492332-9e9582ae-0840-4e3e-a86d-95a01a155505.mp4>

As an additional note, I'd like to share that during implementation, I tried using different tag names like `p` or `span`  but found some issues:
- `p`: On Android, it doesn't allow to add new lines at the end of the content.
https://github.com/WordPress/gutenberg/blob/51db4bf888e6b18cf9d18266108114d61070f3ad/packages/rich-text/src/component/index.native.js#L1095-L1097
- `span`: Automatically wraps the `span` tags with `p` tags. Besides, adding new lines leads to odd issues where lines are merged or add extra newlines 🙃.

geriux replied:
Sounds good! At first I was wondering how it was showing the monospace font without the `pre` tag but I checked we are setting the font from the styles so we are good 👍

geriux commented:
This is working well! Really nice approach to start using RichText in the PlainText component 🤩 I only left one comment, let me know what you think.

fluiddot committed:
Remove tagName prop from PlainText component

fluiddot commented:
> This is working well! Really nice approach to start using RichText in the PlainText component 🤩 I only left one comment, let me know what you think.

Thanks for the review @geriux ! I've just addressed the comment in the latest commit https://github.com/WordPress/gutenberg/pull/41155/commits/0433a9274922d8f090286314a84fc2e0bbc3f3ef. As described [here](https://github.com/WordPress/gutenberg/pull/41155#discussion_r877903926), I decided to remove the `tagName` prop from the render function, so now it can be passed from the parent. If it's not defined, it will use the default value `div`.

geriux reviewed (approved):
LGTM! I tested this on both iOS and Android and it's working well! Nice work! Let's see how this goes but hopefully, we can do the same for other blocks using PlainText. 🚀
